### ISSUE TITLE
Error during seeding (with Uuid)

### ISSUE THREAD

**Issue by xcalibur1978** [spatie/laravel-permission#1478]
I went through [the Uuid setup](https://docs.spatie.be/laravel-permission/v3/advanced-usage/uuid/) and added a seeder which is setup like the example [here](https://docs.spatie.be/laravel-permission/v3/basic-usage/new-app/#add-some-basic-permissions)

I can run the migrations without any porblems. All fields are created as a CHAR 36 field (nullable).

When i try to run the seeder, the following error gets thrown:

> Object of class Ramsey\Uuid\Rfc4122\UuidV4 could not be converted to int

This error occurs when i run 
> Spatie\Permission\Models\Role::givePermissionTo()

Can anyone help me to find out what causes this issue?

In my migrations i use the following lines to add the 'id' field: 

> $table->uuid('id');
> $table->primary('id');

In the seeder I use:
> use Illuminate\Support\Str;
> Permission::create(['name' => 'create objects', 'id' => Str::uuid()->toString()]);
> $role1 = Role::create(['name' => 'admin', 'id' => Str::uuid()->toString()]);
> $role1->givePermissionTo('create objects');

**Comment by drbyte** [comment]
> > Object of class Ramsey\Uuid\Rfc4122\UuidV4 could not be converted to int

Is it possible you haven't set `$incrementing` to false and `$keyType` to string?

**Comment by xcalibur1978** [comment]
I am not really familiar with using traits. I tried two ways to apply the $incrementing en $keyType settings: 

**Method 1** 
<?php

namespace App;

use Illuminate\Database\Eloquent\Model;
use Spatie\Permission\Models\Role as SpatieRole;

Extending the Spatie Role (and the persmission) model: 

> use Spatie\Permission\Models\Role as SpatieRole;
> class Role extends SpatieRole
> {
>     protected $keyType = 'string';
>     public $incrementing = false;
> }

**Method 2**
Adding a Trait (however i don't know where and how I can apply the Trait).

> <?php
namespace App;

use Facades\Str;

trait UuidTrait
{
    public $incrementing = false;
    protected $keyType = 'string';

    protected static function boot()
    {
        parent::boot();

        static::creating(function ($model) {
            $model->{$model->getKeyName()} = $model->{$model->getKeyName()} ?: Str::uuid()->toString();
        });
    }
}

**Comment by drbyte** [comment]
Traits are just like running `include TraitFileName.php;` in procedural code: in a class they let you define methods/properties on that class just by `use`-ing the trait. That way you maintain all that common code in one place. This is how the `HasRoles` trait works on this package. It's also how Laravel's `RefreshDatabase` feature works in `tests`.

If you go with your method 1 then you'll probably also want to copy in the `boot` method from the trait as well.

If you go with your method 2 then inside your \App\Role you can just `use UuidTrait;` the same way you include the `use HasRoles;` trait in your `User` model.

In fact ...
QUESTION: why are you doing uuid in the first place? 
If you're using uuid in your User models, haven't you also implemented these uuid things there? ($incrementing, $keyType, and boot method to set key to uuid on creating).

**Comment by xcalibur1978** [comment]
@drbyte, thanks for the explanation.  It helped me enough to get me back on the right track again. 
In my models I added the attributes $incrementing = false and protected $keyTpe = 'string'. However, I did not use the Trait. Thus, the boot part of the trait was never executed in my models. 

First i added the trait (without removing the attributes.  It resulted in the following error:

>  Illuminate\Database\Eloquent\Model and App\UuidTrait define the same property ($incrementing) in the composition of App\User. However, the definition differs and is considered incompatible. Class was composed

After removing the attributes from the models and sololy using the trait, everything works like a charm.  

> public $incrementing = false;
>  protected $keyType = 'string';
